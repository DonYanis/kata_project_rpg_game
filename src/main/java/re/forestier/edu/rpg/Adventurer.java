package re.forestier.edu.rpg;

import java.util.HashMap;
import java.util.Map;

public class Adventurer implements AvatarClass {
    @Override
    public String getName() {
        return "ADVENTURER";
    }

    @Override
    public Map<Integer, Map<String, Integer>> getAbilitiesPerLevel() {
        Map<Integer, Map<String, Integer>> abilities = new HashMap<>();
        abilities.put(1, Map.of("INT", 1, "DEF", 1, "ATK", 3, "CHA", 2));
        abilities.put(2, Map.of("INT", 2, "CHA", 3));
        abilities.put(3, Map.of("ATK", 5, "ALC", 1));
        abilities.put(4, Map.of("DEF", 3));
        abilities.put(5, Map.of("VIS", 1, "DEF", 4));
        return abilities;
    }

    @Override
    public void applyHealthBonus(Player player) {
        player.currentHealthPoints += 2;
        if (player.retrieveLevel() < 3) {
            player.currentHealthPoints -= 1;
        }
    }
}
