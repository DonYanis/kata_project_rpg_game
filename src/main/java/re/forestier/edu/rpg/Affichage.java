package re.forestier.edu.rpg;

public class Affichage {

    public static String afficherJoueur(player player) {
        final String[] finalString = {"Joueur " + player.Avatar_name + " joué par " + player.playerName};
        finalString[0] += "\nNiveau : " + player.retrieveLevel() + " (XP totale : " + player.xp + ")";
        finalString[0] += "\n\nCapacités :";
        player.abilities.forEach((name, level) -> {
            finalString[0] += "\n   " + name + " : " + level;
        });
        finalString[0] += "\n\nInventaire :";
        player.inventory.forEach(item -> {
            finalString[0] += "\n   " + item;
        });

        return finalString[0];
    }
}
