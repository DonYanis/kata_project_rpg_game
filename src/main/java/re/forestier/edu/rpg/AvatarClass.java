package re.forestier.edu.rpg;

import java.util.Map;

public interface AvatarClass {
    String getName();
    Map <Integer, Map<String, Integer>> getAbilitiesPerLevel();
    void applyHealthBonus(player player);
}
