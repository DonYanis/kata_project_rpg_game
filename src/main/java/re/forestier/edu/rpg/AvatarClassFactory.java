package re.forestier.edu.rpg;

import java.util.Map;

public class AvatarClassFactory {
    
    private static final Map<String, AvatarClass> classMap = Map.of(
        "ADVENTURER", new Adventurer(),
        "ARCHER", new Archer(),
        "DWARF", new Dwarf()
    );

    public static AvatarClass getAvatarClass(String className) {
        if (!classMap.containsKey(className)) {
            throw new IllegalArgumentException("Unknown class: " + className);
        }
        return classMap.get(className);
    }
}
