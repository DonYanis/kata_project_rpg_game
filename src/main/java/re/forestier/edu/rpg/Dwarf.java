package re.forestier.edu.rpg;

import java.util.HashMap;
import java.util.Map;

public class Dwarf implements AvatarClass {

    @Override
    public String getName() {
        return "DWARF";
    }

    @Override
    public Map<Integer, Map<String, Integer>> getAbilitiesPerLevel() {
        Map<Integer, Map<String, Integer>> abilities = new HashMap<>();
        abilities.put(1, Map.of("ALC", 4, "INT", 1, "ATK", 3));
        abilities.put(2, Map.of("DEF", 1, "ALC", 5));
        abilities.put(3, Map.of("ATK", 4));
        abilities.put(4, Map.of("DEF", 2));
        abilities.put(5, Map.of("CHA", 1));
        return abilities;
    }

    @Override
    public void applyHealthBonus(player player) {
        if(player.inventory.contains("Holy Elixir")) {
            player.currenthealthpoints += 1;
        }
        player.currenthealthpoints += 1;
    }
}
