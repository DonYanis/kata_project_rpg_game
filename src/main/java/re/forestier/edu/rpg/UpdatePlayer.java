package re.forestier.edu.rpg;

import java.util.HashMap;
import java.util.Random;

public class UpdatePlayer {

    private final static String[] objectList = {"Lookout Ring : Prevents surprise attacks","Scroll of Stupidity : INT-2 when applied to an enemy", "Draupnir : Increases XP gained by 100%", "Magic Charm : Magic +10 for 5 rounds", "Rune Staff of Curse : May burn your ennemies... Or yourself. Who knows?", "Combat Edge : Well, that's an edge", "Holy Elixir : Recover your HP"
    };

    public static HashMap<String, HashMap<Integer, HashMap<String, Integer>>> abilitiesPerTypeAndLevel() {
        HashMap<String, HashMap<Integer, HashMap<String, Integer>>> abilitiesPerTypeAndLevel = new HashMap<>();

        HashMap<Integer, HashMap<String, Integer>> adventurerMap = new HashMap<>();
        HashMap<String, Integer> adventurerLevel1 = new HashMap<>();
        adventurerLevel1.put("INT", 1);
        adventurerLevel1.put("DEF", 1);
        adventurerLevel1.put("ATK", 3);
        adventurerLevel1.put("CHA", 2);
        adventurerMap.put(1, adventurerLevel1);

        HashMap<String, Integer> adventurerLevel2 = new HashMap<>();
        adventurerLevel1.put("INT", 2);
        adventurerLevel1.put("CHA", 3);
        adventurerMap.put(2, adventurerLevel2);

        HashMap<String, Integer> adventurerLevel3 = new HashMap<>();
        adventurerLevel3.put("ATK", 5);
        adventurerLevel3.put("ALC", 1);
        adventurerMap.put(3, adventurerLevel3);

        HashMap<String, Integer> adventurerLevel4 = new HashMap<>();
        adventurerLevel4.put("DEF", 3);
        adventurerMap.put(4, adventurerLevel4);

        HashMap<String, Integer> adventurerLevel5 = new HashMap<>();
        adventurerLevel5.put("VIS", 1);
        adventurerLevel5.put("DEF", 4);
        adventurerMap.put(5, adventurerLevel5);

        abilitiesPerTypeAndLevel.put("ADVENTURER", adventurerMap);


        HashMap<Integer, HashMap<String, Integer>> archerMap = new HashMap<>();
        HashMap<String, Integer> archerLevel1 = new HashMap<>();
        archerLevel1.put("INT", 1);
        archerLevel1.put("ATK", 3);
        archerLevel1.put("CHA", 1);
        archerLevel1.put("VIS", 3);
        archerMap.put(1, archerLevel1);

        HashMap<String, Integer> archerLevel2 = new HashMap<>();
        archerLevel2.put("DEF", 1);
        archerLevel2.put("CHA", 2);
        archerMap.put(2, archerLevel2);

        HashMap<String, Integer> archerLevel3 = new HashMap<>();
        archerLevel3.put("ATK", 3);
        archerMap.put(3, archerLevel3);

        HashMap<String, Integer> archerLevel4 = new HashMap<>();
        archerLevel4.put("DEF", 2);
        archerMap.put(4, archerLevel4);

        HashMap<String, Integer> archerLevel5 = new HashMap<>();
        archerLevel5.put("ATK", 4);
        archerMap.put(5, archerLevel5);

        abilitiesPerTypeAndLevel.put("ARCHER", archerMap);


        HashMap<Integer, HashMap<String, Integer>> dwarf = new HashMap<>();
        HashMap<String, Integer> dwarfLevel1 = new HashMap<>();
        dwarfLevel1.put("ALC", 4);
        dwarfLevel1.put("INT", 1);
        dwarfLevel1.put("ATK", 3);
        dwarf.put(1, dwarfLevel1);

        HashMap<String, Integer> dwarfLevel2 = new HashMap<>();
        dwarfLevel2.put("DEF", 1);
        dwarfLevel2.put("ALC", 5);
        dwarf.put(2, dwarfLevel2);

        HashMap<String, Integer> dwarfLevel3 = new HashMap<>();
        dwarfLevel3.put("ATK", 4);
        dwarf.put(3, dwarfLevel3);

        HashMap<String, Integer> dwarfLevel4 = new HashMap<>();
        dwarfLevel4.put("DEF", 2);
        dwarf.put(4, dwarfLevel4);

        HashMap<String, Integer> dwarfLevel5 = new HashMap<>();
        dwarfLevel5.put("CHA", 1);
        dwarf.put(5, dwarfLevel5);

        abilitiesPerTypeAndLevel.put("DWARF", dwarf);

        return abilitiesPerTypeAndLevel;
    }

    public static boolean addXp(player player, int xp) {
        int currentLevel = player.retrieveLevel();
        player.xp += xp;
        int newLevel = player.retrieveLevel();

        if (newLevel != currentLevel) {
            // Player leveled-up!
            // Give a random object
            ;
            Random random = new Random();
            player.inventory.add(objectList[random.nextInt(objectList.length - 0) + 0]);

            // Add/upgrade abilities to player
            HashMap<String, Integer> abilities = abilitiesPerTypeAndLevel().get(player.getAvatarClass()).get(newLevel);
            abilities.forEach((ability, level) -> {
                player.abilities.put(ability, abilities.get(ability));
            });
            return true;
        }
        return false;
    }

    // majFinDeTour met Ã  jour les points de vie
    public static void majFinDeTour(player player) {
        if(player.currenthealthpoints == 0) {
            System.out.println("Le joueur est KO !");
            return;
        }

        if(player.currenthealthpoints < player.healthpoints/2) {
            if(!player.getAvatarClass().equals("ADVENTURER")) {
                if(player.getAvatarClass().equals("DWARF")) {
                    if(player.inventory.contains("Holy Elixir")) {
                        player.currenthealthpoints+=1;
                    }
                    player.currenthealthpoints+=1;
                } else if(player.getAvatarClass().equals("ADVENTURER")) {
                    player.currenthealthpoints+=2;
                }


                if(player.getAvatarClass().equals("ARCHER")) {
                    player.currenthealthpoints+=1;
                    if(player.inventory.contains("Magic Bow")) {
                        player.currenthealthpoints+=player.currenthealthpoints/8-1;
                    }
                }
            } else {
                player.currenthealthpoints+=2;
                if(player.retrieveLevel() < 3) {
                    player.currenthealthpoints-=1;
                }
            }
        } else if(player.currenthealthpoints >= player.healthpoints/2){
            if(player.currenthealthpoints >= player.healthpoints) {
                player.currenthealthpoints = player.healthpoints;
                return;
            }
        }


        if(player.currenthealthpoints >= player.healthpoints) {
            player.currenthealthpoints = player.healthpoints;
        }
    }
}