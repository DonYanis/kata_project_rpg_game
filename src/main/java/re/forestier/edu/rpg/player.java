package re.forestier.edu.rpg;

import java.util.ArrayList;
import java.util.HashMap;

public class player {
    public String playerName;
    public String Avatar_name;
    private String AvatarClass;

    public Integer money;
    private Float __real_money__;


    public int level;
    public int healthpoints;
    public int currenthealthpoints;
    protected int xp;


    public HashMap<String, Integer> abilities;
    public ArrayList<String> inventory;
    public player(String playerName, String avatar_name, String avatarClass, int money, ArrayList<String> inventory) {
        if (!avatarClass.equals("ARCHER") && !avatarClass.equals("ADVENTURER") && !avatarClass.equals("DWARF") ) {
            return;
        }

        this.playerName = playerName;
        Avatar_name = avatar_name;
        AvatarClass = avatarClass;
        this.money = Integer.valueOf(money);
        this.inventory = inventory;
        this.abilities = UpdatePlayer.abilitiesPerTypeAndLevel().get(AvatarClass).get(1);
    }

    public String getAvatarClass () {
        return AvatarClass;
    }

    public void removeMoney(int amount) throws IllegalArgumentException {
        if (money - amount < 0) {
            throw new IllegalArgumentException("Player can't have a negative money!");
        }

        money = Integer.parseInt(money.toString()) - amount;
    }
    public void addMoney(int amount) {
        var value = Integer.valueOf(amount);
        money = money + (value != null ? value : 0);
    }
    public int retrieveLevel() {
        // (lvl-1) * 10 + round((lvl * xplvl-1)/4)
        HashMap<Integer, Integer> levels = new HashMap<>();
        levels.put(2,10); // 1*10 + ((2*0)/4)
        levels.put(3,27); // 2*10 + ((3*10)/4)
        levels.put(4,57); // 3*10 + ((4*27)/4)
        levels.put(5,111); // 4*10 + ((5*57)/4)
        //TODO : ajouter les prochains niveaux

        if (xp < levels.get(2)) {
            return 1;
        }
        else if (xp < levels.get(3)) {return 2;
        }
        if (xp < levels.get(4)) {
            return 3;
        }
        if (xp < levels.get(5)) return 4;
        return 5;
    }

    public int getXp() {
        return this.xp;
    }

    /*
    Ингредиенты:
        Для теста:

            250 г муки
            125 г сливочного масла (холодное)
            70 г сахара
            1 яйцо
            1 щепотка соли
     */

}