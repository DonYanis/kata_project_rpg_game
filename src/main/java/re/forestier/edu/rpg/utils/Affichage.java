package re.forestier.edu.rpg.utils;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import re.forestier.edu.rpg.Player;
import re.forestier.edu.rpg.object.Object;

public class Affichage {

    private static final List<String> ABILITY_ORDER = Arrays.asList("DEF", "VIS", "ATK", "CHA", "INT", "ALC");

    public static String afficherJoueur(Player player) {
        StringBuilder finalString = new StringBuilder();

        appendPlayerDetails(finalString, player);
        appendAbilities(finalString, player);
        appendInventory(finalString, player);

        return finalString.toString();
    }

    public static String printPlayerInMarkDown(Player player) {
        StringBuilder profile = new StringBuilder();

        appendHeader(profile, player);
        appendPlayerStatus(profile, player);
        appendStats(profile, player);
        appendInventoryStatus(profile, player);
        appendFooter(profile);

        return profile.toString();
    }

    
    private static void appendPlayerDetails(StringBuilder finalString, Player player) {
        finalString.append(String.format("Joueur %s joué par %s", player.getAvatarName(), player.getPlayerName()))
                .append("\nNiveau : ").append(player.retrieveLevel())
                .append(" (XP totale : ").append(player.getXp()).append(")\n\nCapacités :");
    }


    private static void appendAbilities(StringBuilder finalString, Player player) {
        ABILITY_ORDER.stream()
                .filter(ability -> player.getAbilities().containsKey(ability)) 
                .forEach(ability -> finalString.append("\n   ").append(ability).append(" : ").append(player.getAbilities().get(ability)));
    }


    private static void appendInventory(StringBuilder finalString, Player player) {
        finalString.append("\n\nInventaire :");
        player.getInventory().getObjects().forEach(item -> {
            finalString.append("\n   ").append(((Object) item).getName());
    });
    }

    private static void appendHeader(StringBuilder profile, Player player) {
        profile.append("## ============= PLAYER PROFILE =============  \n")
                .append("**+ Name:** ").append(player.getPlayerName()).append("  \n")
                .append("**+ Controlled By:** ").append(player.getAvatarName()).append("  \n")
                .append("**+ Class:** ").append(player.getAvatarClass()).append("  \n")
                .append("# -------------------------------------------  \n");
    }

    private static void appendPlayerStatus(StringBuilder profile, Player player) {
        profile.append("**+ LEVEL:** ").append(player.getLevel()).append("  \n")
                .append("**+ MONEY:** ").append(player.getMoney()).append(" Coins  \n")
                .append("**+ XP:** ").append(player.getXp()).append("  \n")
                .append("**+ HEALTH:** ").append(player.getCurrentHealthPoints()).append("/")
                .append(player.getHealthPoints()).append(" HP  \n")
                .append("# -------------------------------------------  \n");
    }

    private static void appendStats(StringBuilder profile, Player player) {
        profile.append("**+ STATS:**  \n");
        Map<String, Integer> stats = player.getAbilities();
        stats.forEach((key, value) -> {
            profile.append("- **[").append(key).append("]** : ").append(value).append("  \n");
        });
        profile.append("# -------------------------------------------  \n");
    }

    private static void appendInventoryStatus(StringBuilder profile, Player player) {
        profile.append("**+ INVENTORY STATUS:** ").append(player.getInventory().getTotalWeight())
                .append("/").append(player.getMaxWeight()).append(" kg  \n")
                .append("**+ ITEMS HELD:**  \n");
        player.getInventory().getObjects().forEach(item -> {
            profile.append("- ").append(item.getName()).append("  \n");
        });
    }

    private static void appendFooter(StringBuilder profile) {
        profile.append("## =========================================  \n");
    }
}
