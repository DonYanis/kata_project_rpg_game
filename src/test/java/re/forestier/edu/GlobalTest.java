package re.forestier.edu;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import re.forestier.edu.rpg.UpdatePlayer;
import re.forestier.edu.rpg.utils.Display;
import re.forestier.edu.rpg.Player;

import java.util.ArrayList;

import static org.approvaltests.Approvals.verify;

public class GlobalTest {

    @Test
    @DisplayName("test print player profile")
    void testDisplayBase() {
        Player player = new Player("Florian", "Gnognak le Barbare", "ADVENTURER", 200, new ArrayList<>(),5);
        UpdatePlayer.addXp(player, 20);
        player.clearInventory();

        verify(Display.printPlayer(player));
    }

    @Test
    @DisplayName("test print player profile in MarkDown")
    void testDisplayMarkdown() {
        Player player = new Player("Florian", "Gnognak le Barbare", "ADVENTURER", 200, new ArrayList<>(),5);
        UpdatePlayer.addXp(player, 5);
        player.addObject("Magic Bow");
        player.setHealthPoints(100);
        player.setCurrentHealthPoints(70);

        verify(Display.printPlayerInMarkDown(player));
    }
}
