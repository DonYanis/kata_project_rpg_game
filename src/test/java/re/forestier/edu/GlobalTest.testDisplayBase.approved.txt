Joueur Gnognak le Barbare joué par Florian
Niveau : 2 (XP totale : 20)

Capacités :
   DEF : 1
   ATK : 3
   CHA : 3
   INT : 2

Inventaire :